### PR TITLE
PERF: Allow ensure_index_from_sequence to return RangeIndex

### PR DESCRIPTION
**Pull request by mroeschke**
Discovered in https://github.com/pandas-dev/pandas/pull/57441

Builds on https://github.com/pandas-dev/pandas/pull/57752

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -7156,6 +7155,43 @@ def shape(self) -> Shape:
...
+    Parameters
+    ----------
+    sequence : 1D sequence
+    names : sequence of str
+
+    Returns
+    -------
+    Any : input or range
+    """
+    if hasattr(sequence, "dtype") and not isinstance(sequence, np.ndarray):
```
Comment (WillAyd) [inline pandas/core/indexes/base.py:7173]: I find the duck typing here somewhat non-obvious - maybe comments will help? Is this first branch supposed to be for extension types?

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -7156,6 +7155,43 @@ def shape(self) -> Shape:
...
+        return sequence
+    np_sequence = np.asarray(sequence)
+    if np_sequence.dtype.kind != "i" or len(sequence) == 1:
+        return sequence
+    elif len(sequence) == 0:
+        return range(0)
+    diff = np_sequence[1] - np_sequence[0]
```
Comment (WillAyd) [inline pandas/core/indexes/base.py:7180]: Why are only the first two elements important here?

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/indexes/base.py:7173]: > Is this first branch supposed to be for extension types? 

Essentially yes, I think I can use a more specific check here

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/indexes/base.py:7180]: We're taking an example diff here (just between the first 2 elements) and later seeing if the rest of the diffs between the rest of the elements match

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/indexes/base.py]: Updated the check here to be more explicit (exclude Series and Index objects)

**Comment by mroeschke** [comment]
Any other comments here @WillAyd?

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Nope all good. And thanks for the ping - sorry sometimes fall behind on notifications

**Pull request merged**